### DISCUSSION TITLE
Missing logging information from DS18x20 sensors

### DISCUSSION THREAD

**Discussion by wjohn007** [arendst/Tasmota#20368] [Q&A]
I'm using a WEMOS ESP32 board with Tasmota 13.3.0 and am currently dealing with DS18x20 sensors.
I chose the highest logging level with "Weblog 4" and "Syslog 4".

The sensor data is displayed correctly in the WebUI.

After a restart I would expect log entries from the responsible  module (xsns_05_esp32_ds18x20.ino). 

e.g.
` AddLog(LOG_LEVEL_DEBUG, PSTR(D_LOG_DSB D_SENSORS_FOUND " %d"), DS18X20Data.sensors);`
 
However, this is not the case, not even if I provoke an error (remove sensor).

Is there an explanation for this?

**Comment by barbudor** [comment]
Sensors are initialized early in Tasmota boot process even before the network is up
There is a log buffer that should normally allows to keep log lines until the network is available but this buffer is limited.
So with a high log level which generate many logs, you may loose some early logs from the boot
For `LOG_LEVEL_DEBUG`, level 3 is enough

What are the 1st logs that you she on the console or on syslog ?

On Serial you would be able to get all logs

**Reply by sfromis** [reply]
To verify messages early in the boot process, I usually do a `Restart 1`, and press browser refresh (F5) a few seconds after that, maybe repeated (not fast) to avoid timeouts being at the wrong place in the timing.

However, if the sensor works, you basically already have the info. If not, getting the message can be used to verify that the driver is looking for sensors (it would not do that if no pin configured).

Removal of the sensor does not provoke any error message, only that the count of sensors found will be 0 instead of 1 (or whatever number you have connected).

**Reply by wjohn007** [reply]
Thank you very much for all the answers.
The number of sensors found is logged as follows

```
		00:00:00.273 Project tasmota - Tasmota TestingXX Version 13.3.0(tasmota32)-2_0_14(2023-12-12T14:31:42)
		00:00:00.274 ETH: No ETH MDC and/or ETH MDIO GPIO defined
		00:00:00.300 DSB: Sensors found 1
```
		
After Disconnect of the data-pin

```
	10:58:08.880 DSB: Sensor CRC error
	10:58:08.881 Read sensor 0 failed in Ds18x20EverySecond.
	10:58:08.882 SNS: DS18B20 missed 
```
	

Unfortunately, there is no flag in the telePeriod structure that expresses the validity of the sensor value.
This is set to 0 in the case of an error. But 0 can also be a valid value.
Is there a clear way to detect an error (e.g. in Berry)?

**Reply by sfromis** [reply]
To detect missing sensors, I'd not suggest looking for error messages, but instead have a timer set up for, say, 6 minutes, and restart it every time you get sensor value at TelePeriod (default 5 minutes). Then you can yell if the timer popped, for whatever reason.

**Reply by wjohn007** [reply]
Even if the sensor is missing, the value unfortunately appears as 0 in the teleperiod.

I also found the file xdrv_52_3_berry_onewire.ino, which provides a Onewire driver for Berry.
But I didn't find any documentation for it.

Does this exist?

**Reply by sfromis** [reply]
Did a test with DS18B20 on ESP32, and yanking the data pin out after the sensor was in normal operation. Result was the error messages you posted, but no bad value appearing in the sensor payload. Instead the sensor no longer appeared here, and the whole sensor payload got dropped if I had no other sensors included.

Are you sure that the 0 you see is actually included, along with sensor name and id, in the JSON string Tasmota publishes? I'd call it a bug if this happens.

When it comes to Berry and the `OneWire` class, there is indeed nothing in the docs, but this thread has a bit of description and an example: https://github.com/arendst/Tasmota/pull/12955

Also, available methods:
```
var ow=OneWire(5)
import introspect
introspect.members(ow)
['reset', 'search', 'depower', 'init', 'deinit', 'reset_search', 'skip', '.p', 'write', 'read', 'select', 'target_search']
```
(`init`, `deinit` and `.p` are standard for all, not about using functions of this class)

**Reply by wjohn007** [reply]
I get following loggings

the "all good" scenario
`11:17:17.923 MQT: tele/tasmota_testing/SENSOR = {"Time":"2024-01-03T11:17:17","DS18B20":{"Id":"092220AA1B11","Temperature":20.3},"TempUnit":"C"}`

after yanking the data-pin
`11:19:37.925 MQT: tele/tasmota_testing/SENSOR = {"Time":"2024-01-03T11:19:37","DS18B20":{"Id":"092220AA1B11","Temperature":0.0},"TempUnit":"C"}`

I'm currently working with the OneWire() library. That could be a feasible way.
The comments in the file xdrv_52_3_berry_onewire.ino are helpful.

`  // Berry: `write(byte:int | bytes [, power:bool = false]) -> nil``

The first results seem promising.

```
INFO test.scan - reset search
INFO test.scan - search
INFO test.scan - found supported device :28111BAA20220994
INFO test.every_second - request temperature
INFO test.scan - search
INFO test.scan - finish scanning
INFO test.every_second - will calculate
INFO test.readScratchPad - scratchPad:43014B467FFF0C1079
INFO test.calculateTemperature - neg:0 msb:1 lsb:67 xmsb:2048 xlsb:536 raw:2584 temp:20.1875
INFO test.every_second - request temperature
INFO test.every_second - will calculate
INFO test.readScratchPad - scratchPad:43014B467FFF0C1079
INFO test.calculateTemperature - neg:0 msb:1 lsb:67 xmsb:2048 xlsb:536 raw:2584 temp:20.1875
```

**Comment by wjohn007** [comment]
The OneWire() library works well.
All my requirements can now be met.

![grafik](https://github.com/arendst/Tasmota/assets/115527085/1534ec31-019d-49e4-b74f-5685efa49067)

**Reply by sfromis** [reply]
Great! :smile: 

Your experience of getting a "0" temperature value when yanking the sensor is still what I'd call a bug. Reporting bugs is best when you can provide a repeatable test case. Doing my simple test, where I (as expected) instead lost the sensor from reporting, does indicate that this was not some sort of general issue, but may depend on specific details.
